roachprod: add chaos CLI for failure injection

shailendra-patel opened the pull request:
This commit introduces the `roachprod chaos` command tree, providing CLI support for the failure injection library to test cluster resilience under various failure scenarios.

Changes:
- Add `chaos` root command with global flags for controlling failure duration and cleanup behavior
- Implement network failure commands:
- `chaos network-partition`: bidirectional and asymmetric network splits
- `chaos network-latency`: artificial network delay injection

Example usage:
```shell
roachprod chaos network-partition $CLUSTER --src 1 --dest 2,3 --type asymmetric
```
Epic: none

Release note: None

Fixes: #140634

shailendra-patel committed:
roachprod: add chaos CLI for failure injection

This commit introduces the `roachprod chaos` command tree, providing
CLI support for the failure injection library to test cluster resilience
under various failure scenarios.

Changes:
- Add `chaos` root command with global flags for controlling failure
  duration and cleanup behavior
- Implement network failure commands:
- `chaos network partition bidirectional`: complete network splits
- `chaos network partition asymmetric`: directional traffic dropping
- `chaos network latency`: artificial network delay injection

Example usage:
  roachprod chaos network partition asymmetric mycluster \
    --src 1,2 --dest 3,4 --direction incoming \
    --wait-before-cleanup 10m

Epic: none

Release note: None

DarrylWong reviewed (commented):
First pass looks good overall!

shailendra-patel commented on pkg/cmd/roachprod/cli/chaos.go:
```diff
@@ -0,0 +1,255 @@
+// validateClusterAndNodes validates that:
+// 1. The cluster exists
+// 2. The source and destination nodes are valid for the cluster
+func validateClusterAndNodes(clusterName string, srcNodes, destNodes install.Nodes) error {
```
I tried removing `validateClusterAndNodes ` function. The `GetFailer` do check for cluster, but do not validate the passed node. For example if i run chaos on a cluster with 3 node using following command
```shell
roachprod chaos network partition bidirectional $CLUSTER --src 1 --dst 2,4
```
The failure injection library inject the failure from node 1 to 2 but fails on injection failure from node 1 to 4.
This leaves the cluster in bad state. So keeping this check for now.

shailendra-patel commented on pkg/cmd/roachprod/cli/chaos.go:
```diff
@@ -0,0 +1,255 @@
+		"time to wait before cleaning up the failure")
+	chaosCmd.PersistentFlags().BoolVar(&chaosRunForever,
+		"run-forever", false,
+		"if set, takes precedence over --wait-before-cleanup. On graceful shutdown, cleans up the injected failure")
```
I’d prefer to keep the default behavior as running chaos for a fixed duration using `--wait-before-cleanup`. If someone wants to run chaos indefinitely, they can explicitly use the `--run-forever` flag

DarrylWong commented on pkg/cmd/roachprod/cli/chaos.go:
```diff
@@ -0,0 +1,395 @@
+
+// parseInt32SliceToNodes converts a uint32 slice to install.Nodes
+func parseInt32SliceToNodes(nodes []int32) install.Nodes {
+	if len(nodes) == 0 {
```
nit: this check is unnecessary since the for loop should handle the zero length case

DarrylWong reviewed (commented):
I think using flags is a fine approach for now since we don't exactly know how people will use this tool, but I wonder if it might be worth exploring making the args JSON marshalable (i.e. what chaos mesh does).

Besides that, I took it for a spin and it looks good to me. Just one usability note and a few nits.

shailendra-patel committed:
- flattened command structure to reduce nesting
- created two separate command: network-partition & network-latency
- introduced verbose logging flag

shailendra-patel commented on pkg/cmd/roachprod/cli/chaos.go:
```diff
@@ -0,0 +1,395 @@
+
+// ValidStages returns all valid lifecycle stage values for a failure
+func ValidStages() []string {
+	return []string{
```
As part of the `Inject` stage, I already wait for the failure to propagate before returning. The same applies to `Recover`, which waits until the system stabilizes. Therfore, there is not a need to provide separate options for `WaitForFailure` stages.

shailendra-patel commented on pkg/cmd/roachprod/cli/chaos_network.go:
```diff
@@ -0,0 +1,331 @@
+	var srcNodes, destNodes []int32
+
+	cmd := &cobra.Command{
+		Use:   "bidirectional <cluster>",
```
This is done, made the following changes:
1. Separated out partition and latency as different command.
2. Combined bidirectional and asymmetric network partition as a single command by using `partitionType` as a flag.

DarrylWong reviewed (approved):
Can you link this issue? https://github.com/cockroachdb/cockroach/issues/140634
